### PR TITLE
delay s3 request until read

### LINKED ISSUE #362: s3 open seek operation try read rest of file into buffer, which makes following read has timeout risk
[Here](https://github.com/RaRe-Technologies/smart_open/blob/master/smart_open/s3.py#L200) makes API call to fetch rest of file into buffer when calling `seek`, which makes `seek` very slow. The API call may be put in `read` method instead.

### PR DESCRIPTION
**Pull request by Gapex**
#### Motivation
- Fixes #362

### CODE REVIEW

**Comment by vuryleo** [comment]
i think adding some test would be better

File: smart_open/s3.py [hunk 1]
```diff
@@ -170,44 +172,58 @@ def __init__(self, s3_object, content_length, version_id=None):
         self._object = s3_object
         self._content_length = content_length
         self._version_id = version_id
+        self._body_reload = True
```
Comment (mpenkov) [inline smart_open/s3.py]: Do you really need this variable?

Isn't is possible to set `self._body = None`, and then expect `self._body is None` to mean the same thing as `self._body_reload is True`?

File: smart_open/s3.py [hunk 2]
```diff
@@ -170,44 +172,58 @@ def __init__(self, s3_object, content_length, version_id=None):
...
+        if self._body_reload:
+            self._load_body()
+
+        try:
+            if size == -1:
+                binary = self._body.read()
+            else:
+                binary = self._body.read(size)
+        except IncompleteReadError:
+            # the underlying connection of streaming body was closed by peer
```
Comment (mpenkov) [inline smart_open/s3.py]: Minor English fixup.

```suggestion
            # The underlying connection of the streaming body was closed by the remote peer.
```

File: smart_open/s3.py [hunk 3]
```diff
@@ -170,44 +172,58 @@ def __init__(self, s3_object, content_length, version_id=None):
...
+            # close the current one explicitly, prepare to open a new and try to read again
```
Comment (mpenkov) [inline smart_open/s3.py]:
```suggestion
            # Close the current connection explicitly, prepare to open a new one and try to read again.
```

File: smart_open/s3.py [hunk 4]
```diff
@@ -170,44 +172,58 @@ def __init__(self, s3_object, content_length, version_id=None):
...
+            self._body.close()
+            self._body_reload = True
+            return self.read(size)
```
Comment (mpenkov) [inline smart_open/s3.py]: You're setting up an infinite recursion here, in theory.

Can you think of a way to break the cycle? We could count the number of consecutive IncompleteReadErrors, and then bail out if we see more than e.g. 10.

File: smart_open/s3.py (hunk 1 above)
Comment (Gapex) [reply smart_open/s3.py]: you're right.

File: smart_open/s3.py (hunk 4 above)
Comment (Gapex) [reply smart_open/s3.py]: What about adding a member field to count the number of retry, when counter less than the max limit, retry, other wise, terminate the action by raising an error?

Comment (mpenkov) [reply smart_open/s3.py]: That's OK, but you need to make sure you're counting _consecutive_ failures, as opposed to total failures. So, you reset the counter to zero on each successful attempt.

**Comment by mpenkov** [comment]
Looks like your tests are failing because you're dereferencing `None`, please have a look at the buildbot output (https://travis-ci.org/RaRe-Technologies/smart_open/jobs/610158039?utm_medium=notification&utm_source=github_status).

File: smart_open/s3.py [hunk 5]
```diff
@@ -199,13 +201,30 @@ def seek(self, position):
...
         else:
             binary = self._body.read(size)
+        return binary
+
+    def read(self, size=-1):
+        if self._position >= self._content_length:
+            return b''
+
+        if self._body is None:
+            # After seek, or the first read, the self._body is None.
```
Comment (vuryleo) [inline smart_open/s3.py]: how about closed?

File: smart_open/s3.py [hunk 6]
```diff
@@ -199,13 +201,30 @@ def seek(self, position):
...
+            self._load_body()
+
+        try:
+            binary = self._read_from_body(size)
+        except IncompleteReadError:
+            # The underlying connection of the streaming body was closed by the remote peer.
+            self._body.close()
```
Comment (vuryleo) [inline smart_open/s3.py]: do we need this?

File: smart_open/s3.py (hunk 5 above)
Comment (Gapex) [reply smart_open/s3.py]: After init, the self._body is None, the connection is not built until the first read.

Comment (vuryleo) [reply smart_open/s3.py]: So you can still read from a `closed` reading stream, interesting. Is this work as designed?

File: smart_open/s3.py (hunk 6 above)
Comment (Gapex) [reply smart_open/s3.py]: I will delete the line.

File: smart_open/s3.py (hunk 5 above)
Comment (Gapex) [reply smart_open/s3.py]: Yes, we could read from a closed stream, even if after the `close()` operation, but there is no such api offer to users. 
In addition, according to the botocore [reference](https://botocore.amazonaws.com/v1/documentation/api/latest/reference/response.html#botocore-response), streaming body doesn't have the `closed` attribute.

**Comment by Gapex** [comment]
@mpenkov , When executing `python setup.py test`, read from a closed streaming body always returns 0 bytes instead of raising an IncompleteReadError, thus the action of reopening an new streaming body had not been triggered. Thus, the SeekableBufferedInputBase thinks the raw reader reaches it's EOF, and just returns the rest data in the buffer.

This makes the unit test  be failed.

**Comment by mpenkov** [comment]
OK, some questions:

- Who returns 0 bytes when reading from a closed stream?
- What depends on the existing behavior (return 0 bytes)?
- The correct behavior is to raise an exception, right?

We probably want to fix this, as it deviates from the behavior of file streams in the stdlib.

**Comment by mpenkov** [comment]
@Gapex What happened here? Why did you close the PR?

**Comment by Gapex** [comment]
Yes, I will open a new pull request.

**Pull request closed without merging**